src/CommonRLInterface.jl

```diff
@@ -4,35 +4,105 @@ using MacroTools
 
 export
     AbstractEnv,
-    step!,
+    AbstractMarkovEnv,
+    AbstractZeroSumEnv,
     reset!,
-    actions
+    actions,
+    observe,
+    act!,
+    player
 
 abstract type AbstractEnv end
 
 """
-    o, r, done, info = step!(env, a)
-
-Advance AbstractEnv `env` forward one step with action `a`.
+An environment that represents an MDP or a POMPD. There is only one player in this environment.
+
+The required interface consists of
+```
+reset!(env)
+actions(env)
+observe(env)
+act!(env, a)
+```
+"""
+abstract type AbstractMarkovEnv <: AbstractEnv end
 
-This function should be provided by every AbstractEnv and should return an observation, reward, Boolean done signal, and any extra information (typically in a NamedTuple).
 """
-function step! end
+An environment that represents a zero-sum two player game, where only one player plays at a time. Player 1 seeks to maximize the reward; player 2 seeks to minimize it.
+
+The required interface consists of
+```
+reset!(env)
+actions(env)
+player(env)
+observe(env)
+act!(env, a)
+```
+"""
+abstract type AbstractZeroSumEnv <: AbstractEnv end
 
 """
-    o = reset!(env)
+    reset!(env::AbstractEnv)
 
-Reset `env` to its initial state and return an initial observation.
+Reset `env` to its initial state and return `nothing`.
+
+This is a *required function* that must be provided by every AbstractEnv.
 """
 function reset! end
 
 """
-    actions(env)
+    actions(env::AbstractEnv)
+
+Return a collection of all of the actions available for AbstractEnv `env`. If the environment has multiple agents, this function should return the union of all of their actions.
+
+This is a *required function* that must be provided by every AbstractEnv.
+
+This function is a *static property* of the environment; the value it returns should not change based on the state.
+
+---
 
-Return a collection of all of the actions available to the agent in AbstractEnv `env`.
+    actions(env::AbstractZeroSumEnv, i::Integer)
+
+Return a collection of all the actions available to player i.
+
+This function is a *static property* of the environment; the value it returns should not change based on the state.
```

Comment:
Maybe we should insist that `valid_action_mask` must be defined with respect to `actions(env)` and not `actions(env, player(env))`.

Also, is there a way to make implementing `actions(env)` mandatory but `actions(env, player)` optional?

Reply:
Yes, currently `actions(env)` is mandatory, but `actions(env, player)` is optional.